### PR TITLE
ARROW-7921: [Go] Add Reset method to various components and clean up comments.

### PR DESCRIPTION
**Pull request by richardartoul**
The reset method allow the data structures to be re-used so they don't have to be allocated over and over again.

### CODE REVIEW

**Comment by wesm** [comment]
Can you please open a JIRA issue for the benefit of our changelog?

File: go/arrow/array/string.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ const (
 	stringArrayMaximumCapacity = math.MaxInt32
 )
 
-// A type which represents an immutable sequence of variable-length UTF-8 strings.
+// String is a type which represents an immutable sequence of variable-length UTF-8 strings.
```
Comment (sbinet) [inline go/arrow/array/string.go]:
```suggestion
// String represents an immutable sequence of variable-length UTF-8 strings.
```

File: go/arrow/array/data.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ import (
 	"github.com/apache/arrow/go/arrow/memory"
 )
 
-// A type which represents the memory and metadata for an Arrow array.
+// Data is a type which represents the memory and metadata for an Arrow array.
```
Comment (sbinet) [inline go/arrow/array/data.go]:
```suggestion
// Data represents the memory and metadata of an Arrow array.
```

File: go/arrow/memory/buffer.go [hunk 3]
```diff
@@ -60,15 +61,28 @@ func (b *Buffer) Release() {
 	}
 }
 
+// Reset resets the buffer for reuse.
+func (b *Buffer) Reset(buf []byte) {
```
Comment (sbinet) [inline go/arrow/memory/buffer.go]: it's a part of the codebase I am not completely fluent with.
but, shouldn't we also somewhat test whether that `Buffer` is `Mutable()` before modifying it?

**Comment by sbinet** [comment]
also, it would be nice to have a few tests for this new API.

File: go/arrow/memory/buffer.go (hunk 3 above)
Comment (richardartoul) [reply go/arrow/memory/buffer.go]: As far as I can tell `mutable` just means `there is a reference to an allocator here so things can be resized`, I don't think it has much bearing on calling `Reset()`.

The `Reset()` is used to reuse data structures so you're also kind of already saying: "I know what I'm doing"

**Comment by richardartoul** [comment]
@sbinet I think I addressed all your feedback. I updated the method strings and added some tests. Not really sure what to do about the mutability thing (I left my thoughts on that) but I don't think it impacts correctness at all and the new methods are mostly optimizations to prevent extra allocations for people who want to eak out some extra perf in the hot path.

**Comment by richardartoul** [comment]
@wesm Opened a ticket!

File: go/arrow/memory/buffer.go (hunk 3 above)
Comment (sbinet) [reply go/arrow/memory/buffer.go]: ok

**Review by sbinet — APPROVED** [review]

**Pull request closed without merging**